### ISSUE TITLE
Allow passing options to sub-lexers.

### ISSUE THREAD

**Issue by jneen** [rouge-ruby/rouge#1381]
**Is your enhancement request related to a problem? Please describe.**
Currently, there is no way to specify rich lexers with options. For example, ERB in a configuration file, with escaped content - we would need to pass `?lang=erb` to the `escape` lexer, and `?parent=conf` to the erb lexer.

**Describe the solution you'd like**
I would propose the syntax `?lang=erb&lang.parent=conf` - there may be a CGI parser out there that can handle it.

**Additional context**
A potential issue is that we may need to specify some keys be inherited (or inherit all keys? that's the default behaviour right now). A good example of this is `debug` - if `?debug=1` is passed to the main lexer, then *all* lexers involved should have debugging turned on.

**Comment by jneen** [comment]
The implementation would happen in `Lexer#lexer_option`, to extract the relevant options from the options hash and pass them into `Lexer#as_lexer`.

**Comment by hugopeixoto** [comment]
I was trying to use the `console` lexer and specifying the `output` option, which itself accepts a `token` option. The documentation says that `output` defaults to `plaintext?token=Generic.Output`, but I'm unable to override it, which I think is the same scenario that you're mentioning.

I got this working by changing `#as_lexer`:

```ruby
# from
when String
  lexer_class = Lexer.find(val)
  lexer_class && lexer_class.new(@options)

# to
when String
  Lexer.find_fancy(val)
```

This allows me to specify the `output.token`:

```
console?output=plaintext%3ftoken=Text
```

Would something like this solve this issue? This no longer passes parent options to descendant lexers, though (lost the `@options`). They probably need to be merged.

**Comment by hugopeixoto** [comment]
I did the change locally and tests pass:
https://github.com/hugopeixoto/rouge/commit/f28eeea9240b0209131631c0c8fce473b96c84a5

`#find_fancy` accepts additional options to be merged, so that solves that issue. This didn't impact the test results, though. I'll see if I can add some tests.

**Comment by jneen** [comment]
That may work for now, but it's important to be able to distinguish between options for the parent lexer and options for other lexers - that's why I wanted to be able to use, in this case, `output=plaintext&output.token=text`.

**Comment by jneen** [comment]
(This will probably have to be a breaking change unfortunately, since we unwisely passed the flat options everywhere...)

**Comment by hugopeixoto** [comment]
> That may work for now, but it's important to be able to distinguish between options for the parent lexer and options for other lexers - that's why I wanted to be able to use, in this case, `output=plaintext&output.token=text`.

I agree. The patch I shared allows you to do something similar to your suggestion, but with a different syntax:

```
console?output=plaintext%3ftoken=Text
```

The `token` param would only get passed to the `output` lexer, not to the `console` one. The `%3f` is an escaped `?`, making the `output` value `plaintext?token=Text`, which w ould then be picked up by `find_fancy` in `as_lexer`.

I mentioned the `@options` being passed from parent to child only because I was wondering what it's use case is, but your `debug=1` example is a good one where we'd want to propagate some options.

**Comment by jneen** [comment]
oh..... yeah I think I'd rather avoid using the escape syntax if at all possible.

**Comment by hugopeixoto** [comment]
With #1412 merged, can this be closed? Or will you only close the issue when 4.0 gets released?
